### DISCUSSION TITLE
Build step (Typescript included)

### DISCUSSION THREAD

**Discussion by Weredime** [ScratchAddons/ScratchAddons#4886] [Discussion]
Pros of using a build step

1. We can migrate from Vue 1 to some other framework (maybe Svelte, or even Vue 3) for pages
2. We don't have to keep updating dependencies manually (pNPM)
3. CSS modules
4. Types
5. Faster loading of Addons ~(no module.js and comlink)~
6. ...and more

Cons of a build step
> TODO

**Comment by pufferfish101007** [comment]
> Vue 3

Doesn't need a build step if you use render functions (or just boring html)

**Reply by Weredime** [reply]
Fair point, but who wants to use render functions

**Reply by pufferfish101007** [reply]
me :)

**Reply by Weredime** [reply]
it will make the code much more lengthy

**Reply by cobaltt7** [reply]
The problem with Vue 3 is the CSP.

**Comment by WorldLanguages** [comment]
> Faster loading of Addons (no module.js and comlink)

How would a build step remove the need for comlink?

**Reply by WorldLanguages** [reply]
Also, we could use CSS nesting.

**Reply by Weredime** [reply]
For chrome, we don't need comlink, since import() is supported

**Reply by WorldLanguages** [reply]
Using a build step does not make comlink unnecessary.

**Reply by cobaltt7** [reply]
> Also, we could use CSS nesting.

You want that too much :joy:

**Comment by Secret-chest** [comment]
Cons: Much harder to contribute, and I don't want to install TypeScript on my PC.

**Reply by Hans5958** [reply]
Skill issue /s

EDIT: No, I mean, seriously, I would say the investment would worth it, as a former Typescript user.

**Reply by Ascor8522** [reply]
@Secret-chest No, wouldn't be any harder to contribute.

First off, I'd like to say **I understand your concerns**, moving to something new can sometimes be scary. But at the same time, I find it very **bold of you to be against it without having ever used it** or at least tried it. How can you have an opinion on something you don't know anything about? Also, **you aren't giving any valid arguments** to support your claim ("I don't want to" is not a vlid argument).

Anyway, here are a few thing I'd like to point out:

- **If you know JavaScript, then you already know valid TypeScript**. TypeScript just adds features, same as every release of JavaScript comes with new features too. Best part is, **all of those additions are optional**, you are free not to use them if you don't want to.
- All **those features are here to help you**, not change your way of doing things. **It checks for potential problems in your code, that's what it mainly does**. By the way, you **might already be using TypeScript without even realizing it**; opening a JavaScript file VSCode will call the TypeScript language server, same for Intellisense. Those errors you see in your JavaScript, there were made possible by TypeScript checking your code in the background.
- Also, in a good TypeScript codebase, the types shouldn't be a concern. About only the core of the library would have types, and the rest of the code **(all the addons) should in theory just consume them** (use them without writing some).

About installing the compiler, your argument doesn't make any sense.
- Installing the compiler is **exactly the same as installing any JavaScript dependency** from npm, which you most likely have already done if you have ever made a project using NodeJs. But yeah, **needing to run `npm i` once** after cloning the project makes it "much harder to contribute to". Best part, it can be installed as a project dependency, you don't need to install it globally.
- The entier compiler is only one project, it doesn't have any dependencies, and **is only [about 3MB uncompressed](https://bundlephobia.com/package/typescript)**. You can't say the same about all the other tools like ESLint, and the additional rules (packages) you'd usually add to it that this project is recommending using.
- TypeScript would check your code in a much better and deeper way that ESLint does. It checks for the type of every snigle object in your code, whereas ESLint only checks for some common code-style error (valid JS syntax but things probably not intended by the developer).
- Using the TypeScript compiler is event easier; just **call the build script already defined** in the `package.json`. You don't even have to write to write the whole command every time. Even better, if you are using VSCode, just click the task button to call the build script for you. Imagine needing to **click a single button** to compile and type-check the whole code-base, but yeah, that makes it "much harder to contribute to".

**Reply by Secret-chest** [reply]
NPM sucks!

**Reply by Secret-chest** [reply]
And before I could just click reload in the browser extensions page.

**Reply by Weredime** [reply]
> NPM sucks!

use PNPM

**Reply by Ascor8522** [reply]
> NPM sucks!

Again, no arguments here too.

You're not even explaining what part you don't like, and why you don't like it.
What's wrong with npm as a package registry? What's wrong with the npm cli?

> use PNPM

same goes for that too

In what way is it better? What does it seolve?

you should read this https://github.com/ScratchAddons/ScratchAddons/discussions/4886#discussioncomment-3232636

**Reply by Hans5958** [reply]
> NPM sucks!

I'm not sure if it is okay for me to say "Skill issue /s" one more time, or to just stay out of it. In any case, I can see Ascor is being respectable here, and I would do the same to him, and you guys should too.

That being said, with all due respect:

> “Never argue with stupid people, they will drag you down to their level and then beat you with experience.”  
> ― Mark Twain 

> use PNPM

Why? Also, I could argue to use `yarn` but you need a reason why NPM sucks in the first place so I can give you a proper alternative.

> And before I could just click reload in the browser extensions page.

It's 2022 and there are already tools that spoils you with it.

- Pure TypeScript? Configure your `tsconfig.json` and run `tsc --project --watch`
- Webpack? Configure and run [`npx webpack serve`](https://webpack.js.org/configuration/dev-server/).
- Rollup? There is [`--watch`](https://rollupjs.org/guide/en/#-w--watch=) waiting for you.

All of this builds SA as soon as you save, and you can just reload your browser, just like when you're doing JS.

**Comment by Weredime** [comment]
OR, Like @Ascor8522 (rip) originally said in the Discord, we can have a build step for the core only. If we use this system, we keep both the built stuff and core stuff together

**Reply by Ascor8522** [reply]
Separating the core from the addons in two separate projects would be the right way of doing things.

That would mean
- the core could use other (and better) technologies (like, but not limited to TypeScript), while the addons would see no difference (they don't care where the core is coming from, as long as it's there and they know what to expect)
  - by better technologies, I mainly mean using TypeScript, bundling (Rollup seems nice and easy) and minifying (there are Rollup plugins for that) the core (and thus providing typings too) 
- the core can be better versionned, the two projects could move forward independetly from one another; newer versions of the core cvan be published while the addons are still on an older version

But in any case, please, don't put the build output in the repo, like it's the case for [these](https://github.com/ScratchAddons/ScratchAddons/tree/master/libraries/thirdparty) at the time of speaking. Do it the proper way, call them from a CDN like UNPCKG if possible, or use a bundling phase in the project with the addons as well then.

By the way, I know it will be controvertial, but bundling the code from all addons together would help a lot when loading the extension. The more dynamic `import`s you have, the more time is spent parsing JS code and doing requests.

**Reply by Weredime** [reply]
@Ascor8522 Also, why did you leave the discord? lol

**Reply by Ascor8522** [reply]
> why did you leave the discord

Y'all can't be helped.

Here is a more detailed explanation. It's nothing but my point of view. If you aren't happy about it, either ignore it or try improving upon the points I listed.

- [ ] Owner doesn't do much to help the project: can't delegate, prefers to ignore important questions that are core to the project, doesn't have a long term view of what the project should be.
- [ ] Gatekeeper straight rejects any idea he hasn't come up with, ignores any argument not going his way. Doesn't have a mind of his own, sticks to and religiously believes in everything owner said. Might be knowlegeable in some areas but perfers keeping everything to himself. Is also igorant in some other ones but doesn't question himself or the things he made. Is hypocritical, makes claims and acts in does the exact opposite some time later. Doesn't understand people can have a different point of view from his. Only has responsabilities because was there early on in the project.
- [ ] No one knows what this project is supposed to be exactly. A project for kids to have fun submitting code of doubtious quality or a serious project aiming to add reliable features to Scratch? Pick one. Maybe it was made this way in the beginning to attract new contributors, but it clearly hasn't worked, since most contributors already are well known members from the community. Project has never truly transitionned from original author to the org. Decisions taking early on are holding the project back and everybody seems to be fine with it. Are we just going to keep adding addons forever?
- [ ] I've already suggested lots of improvements, every time coming up with lots of arguments supporting my case, but I've only been ignored at best, if not silenced, without prior warning or explanation afterward.
- [ ] Kids like dogs wagging their tails every time owner is online, don't care about anything other than their code being merged and a new release being published. Using "lol" and "lmao" as punctuation.

**Reply by Hans5958** [reply]
> Y'all can't be helped.

Ascor, I can see and understand your doubts on this project, and I'm going to feel pity about it. The way I look is that SA really wanted to be just a casual project, and not something that to be done serious upon.

I think I also have similar viewpoints of yours, but I'm sure I suggested things that just on the large backlog. I think at this time they are focusing on migrating to MV3 and somehow lost focus on some parts of the project. The documentation is now a little bit left behind, it's only someone who told me about it that I can fix it by myself. I guess we got improvements on some QA. We also got debates behind the scenes. In any ways, I know what you feel about this.

That being said, I would also take pity because I sometimes jokingly around on the Discord server. Since the inception of the project, I would take a pledge to be professional on a GitHub, but I'm sure some slipped. I also got into trouble in the Discord server itself. I'm not sure who are you directing specifically, but I know I did at least one of your concerns, and I feel pity about the others as a "maintainer". So, I apologize personally.

**Reply by Ascor8522** [reply]
You're all good, don't worry.

To be honnest, I'm completely fine with SA wanting to be a casual project. Only problem is that it tries to advertise iteself as a serious project. Also the way some members act, by, at the same time considering it at something almost profesionnal and not caring for things most serious projects care about (using the right tools for th job, clear long term view and planning, clear and well defined scope for all features, extensive testing, extensive documentation, complete localization before realease, automation of the error-prone / human-dependant tasks, etc.). That's that kind of hypocritical behavior that has been bugging me for long now.

Also, when the project was only getting started, coming up with a few things, even sketchy,  wasn't a problem, since the goal was just to get started as soon as possible, and have a minimum viable product. But now that this phase has been over for long, things aren't changing, the mindset shifted from "we'll get that fixed later" to "this works, it doesn't need to be changed or improved upon", which I think isn't the right mindset imo.

**Reply by Hans5958** [reply]
To be honest, since the inception of the project, I never seen that this is something professional/serious, at least this is a perspective of me having quite distanced to the core development (as in I don't invest much time on the engine and/or the addons, working on the sidelines). So, I don't have high, high expectations to something as a minimum of proper documentation, even though I'm sure that is a thing that I know is worth it and nice for me.

By the way, I wanted to implement [Danger.js](https://danger.systems/js/) which is nice to automate some checking, and obviously make things easier, but I know for the fact that Apple will be [pissy](https://github.com/ScratchAddons/ScratchAddons/pull/1941#issuecomment-804624046) [mad](https://github.com/ScratchAddons/ScratchAddons/pull/1941#issuecomment-805114223) about these automations. What I have in my mind is to automate, so we can spend more time in other parts of the project, but it makes me a little bit disappointed that it is gone to the direction, and next to none objections because people are so casual. 

And we're not even talking about the other parts of project management. Proper commit messages (`Lint 00000000` is a joke), [proper changelog that shows all the changes](https://github.com/ScratchAddons/ScratchAddons/blob/changelog-detailed/CHANGELOG.md), etc, etc. And then the fact we need to make things easier for people who doesn't know Node.js existed...

I saw similar extensions to SA, and at the same time, wanted for the project to have the highest quality, but I have made my expectations to the project, and I don't think it is a bad way to think about it. I was also thinking if it is better to abstractize to improve the inner workings, discourage select people, or keep being trivial to maintain longetivity for the development, but here we are, producing lower quality product.

Oh, man, the wonders of project management! At least I'm always learning on these projects.

**Reply by lisa-wolfgang** [reply]
> Owner can't delegate

To be fair, steps are being taken to better delegate PR merging. Now two approvals from within the org are all that is required to merge a PR. In addition, we will soon be providing an "approval checklist" to make sure that the PR follows given quality standards, which are listed here: https://github.com/ScratchAddons/website-v2/pull/219

**Comment by Ascor8522** [comment]
Agree

Would allow the project to use better tooling, mainly talking about

- [ ] TypeScript
  - the core is a mess, we need more type safety
- [ ] Dependencies from npm
  - please stop adding built files to version control, that's not how git should be used, this is only but an ugly hack as of now
  - would allow keeping our deps up to date more easily
- [ ] use a bundler
  - that solve the problem created above with npm deps
  - as well as bundling all the core and addons together, dynamic imports are only great for when developing, not for production.
  - That would allow all the addons and the extension in general to load faster.
- [ ] use other code checking tools, officially to the project
  - like linting and formatting, git hooks, etc.
  - As of now, it's opt-in, but a build step could take care of it automatically by pressing a single button.
- [ ] SCSS, basically the same point as TypeScript + bundling, but for styles
- [ ] documentation (mainly TSDoc and TypeDoc)
  - technical doc should be extracted from the code-base.
  - It's not possible as of now, mainly because there isn't enough doc in the code, or it is incomplete / outdated.
  - This would allow the website to be build automatically when the code changes, to reflect those changes.
  - This would also allow for a better build system for the website, but that's another topic.
- [ ] finally, I'd suggest switching to Preact.
  - Works great with TypeScript (can't say the same about Vue);
  - allows us to use TSX, a better syntax that creating DOM nodes by hand in the addons, same for the extension part;
  - compiles to plain JavaScript;
  - bundle isn't as minimal as Svelte but source is closer to plain JavaScript;
  - can even be used without TSX and can call functions to create the UI programmatically, whereas it's not possible Svelte;
  - is much more popular and has better resources;
  - can easily be integrated with the Scratch website wich uses React thanks to a compat module;
  - comes as ES modules and can be bundled and treeshaken easily.

**Reply by Weredime** [reply]
> * finally, I'd suggest switching to Preact.

Why not `dom-chef`? It turns JSX/TSX into pure DOM elements.

**Reply by Ascor8522** [reply]
It's the same thing, the TS compiler (or Babel) can turn the JSX/TSX call into a series of function calls. Those function being from the library of your choice: react, preact, or dom-chef. They moreless all work in the same way.

The reason I recommended Preact is mainly
- because of its size (3kB for Preact vs "7kB" for React),
- the fact that it using ES modules natively (whereas React doesn't),
- has an extensive TypeScript support
- that it can integrate nicely with React (thanks to its compat module)
- that it is has a relativerly large community

Btw, I didn't know about dom-chef, but I wouldn't be against it if it had some clear advantages that the other main lib don't.

**Reply by Hans5958** [reply]
TypeScript: My point still stands from the last issue (#1301). Easier for me to call the types, and more type-safe programming. Although it's been a while for me to code the addons, and JSDoc works in some way?

npm: I agree based on the fact that we need to use dependencies.

Bundler: I never used specalized tools such as Webpack, Grunt, etc, but based on my experience on using Parcel.js, this does not hurt to try. You would also got lighter distributables. Don't forget you can also do polyfill to stop worrying with the new-fangled features.

Other checking tools: I only tried git hooks once and it is a surprise for me, I think it is good if you really, really wanted to maintain code quality, but I have little doubts on it. I am neutral on this.

SCSS: Been a supporter since the inception of the SA website, and it was worth it.

Docs: Again, someone already tried to use JSDoc, but it is deprecated in favor of manual documentation. Search `jsdoc` on the issue board. I would, however, look forward for the second try.

Preact: I am neutral on these web frameworks since I don't have enough experience on it.

**Comment by Hans5958** [comment]
Related: #1301

**Comment by Hans5958** [comment]
**Update**: A build script is now proposed for the settings page. #6049

**Reply by Hans5958** [reply]
Focus seemed to keep Node.js away. https://github.com/ScratchAddons/ScratchAddons/pull/6049#discussion_r1196631694

**Comment by Hans5958** [comment]
I think I have wrote this but I'm unsure, so I will write a new one here, with a fresh point of view.

As an extension with close to 200k users, I don't think we should keep this concept of "no build scripts" just for the sake of appealing those who don't care enough about Node.js.

As an established browser extension that is big enough without any promotion inside the supported service itself, the focus for developer experience (e.g. "you can build this without any Node.js") shouldn't be the primary goal, an instead should be the user experience. 

This does not mean the developer experience should be played down. This also doesn't mean I want to be such a gatekeeper for the SA's development. It should be important enough for the longevity of the project, yes, but I don't think we should bring down the quality of the project [^1] just in favor of "they will be disappointed if Node.js is used!"

I get it, "I don't want to get out from the safe zone!" or "It's just how it works" But, I don't think we can "improve" further if we limit ourselves to appeal the lower-end kind of developers.

PS. Please, upvote instead of :+1:.

[^1]: for example:
  user facing: #2531
  dev facing: various bodge such as libraries, different manifest, etc

**Reply by Ascor8522** [reply]
Tho I'm no longer involved in this project, I'm happy this topic is finally addressed, and I fully agree with the point made.
Also wanted to say that I can  always provide help or tips if you decide to go down the TypeScript / ESBuild / Webpack / build script road.

**Reply by Secret-chest** [reply]
I don't want that rubbish on my PC if I only want to develop for SA! It conflicts with APT!

**Reply by Samq64** [reply]
We're using a much newer version than Scratch and you don't have to install npm packages globally.

**Reply by GrahamSH-LLK** [reply]
I completely agree. My PR works the way it does because that's been the status quo.

**Reply by DNin01** [reply]
This would also be a good challenge for any of us who haven't ever used Node.js or npm before. It seems like there's only a few things to learn so it's probably easier than it seems, as long as you accept and adapt to the change.

**Reply by DNin01** [reply]
When we merge the upgrade to Vue3, I would appreciate a more in-depth guide or a link to one, though. It'd be helpful to know more than just the commands to install dependencies and compile.

**Reply by Hans5958** [reply]
> Tho I'm no longer involved in this project, I'm happy this topic is finally addressed, and I fully agree with the point made. Also wanted to say that I can always provide help or tips if you decide to go down the TypeScript / ESBuild / Webpack / build script road.

I didn't expect you to return on this discussion, but I will appreciate the help if we get these things along. I still remember what you said on this thread months ago, and my point today is still the same as last time.

> This would also be a good challenge for any of us who haven't ever used Node.js or npm before. It seems like there's only a few things to learn so it's probably easier than it seems, as long as you accept and adapt to the change.
>
> When we merge the upgrade to Vue3, I would appreciate a more in-depth guide or a link to one, though. It'd be helpful to know more than just the commands to install dependencies and compile.

As of my experience it should be painless enough. Rest assure the docs would be as useful as what I did on the website, or more, that is if we take more time on the documentation in general.

----

And now, the elephant of the room.

> I don't want that rubbish on my PC if I only want to develop for SA! It conflicts with APT!

It is quite appalling that you are still as ignorant as few months ago, though I still appreciate you are still around for me to address it, appropriately in the post about the motivation for one.

How does `apt` play on this problem? Sure, you have to use Node.js, but all dependencies will be installed in a per-project basis (yes, [this is being ridiculed](https://www.reddit.com/r/ProgrammerHumor/comments/6s0wov/heaviest_objects_in_the_universe/), but there are solutions such as `pnpm` that can partially solve this, which you have mentioned). This is not like Python which, yes, installs globally, though `npm` packages can be installed globally, in no way this not the default option.

Also, what is this "rubbish" you are talking about? I didn't want to dig deeper on your history, but as a programmer myself I have **multiple** software development kits for **multiple** languages (for example, I have Ruby, Node.js, Python, and even Scratch!). I'm not sure why adding Node.js on your environment would affect you in a way you called it "rubbish", looking that you at least have a GitHub account.[^1]

If even installing or using Node.js is your problem, perhaps you are the part of the problem as I wrote.

[^1]: This is my safe assessment, but if it isn't, I apologize for being presumptuous.

**Reply by lisa-wolfgang** [reply]
Thought it would be worth noting that GitHub now offers Codespaces for free for a fairly generous amount of time per month. That could be used for development if there are qualms about installation.

**Reply by GrahamSH-LLK** [reply]
> Thought it would be worth noting that GitHub now offers Codespaces for free for a fairly generous amount of time per month. That could be used for development if there are qualms about installation.

Well I'm not quite sure how you could test it but yeah

**Reply by Secret-chest** [reply]
I don't want containers. Yes, I use pip since it plays nicely with debian.

**Reply by GrahamSH-LLK** [reply]
> I don't want containers. Yes, I use pip since it plays nicely with debian.

There wouldn't be any containers. I'm not sure where you're coming from, but you're making this much bigger of a deal than it is.

**Reply by Hans5958** [reply]
> > Thought it would be worth noting that GitHub now offers Codespaces for free for a fairly generous amount of time per month. That could be used for development if there are qualms about installation.
> 
> Well I'm not quite sure how you could test it but yeah 

Maybe they could pack it inside Codespace and download it so they can install it on their browsers. Would be tedious but that's possible.

**Reply by Secret-chest** [reply]
One of my problems is that VS Code sucks (very weird), any IntelliJ with JS support is paid, Atom is gone...

**Reply by Hans5958** [reply]
That is nonsense. Node.js support is literally [built-in](https://code.visualstudio.com/docs/nodejs/nodejs-tutorial), so the tool can't be blamed.

Also, I use exclusively VS Code for my Node.js projects, and I haven't got any problems. I never use IntelliJ other than Java, and Atom is discontinued.

**Reply by Secret-chest** [reply]
I thought I already said that this stupid npm conflicts with GPU drivers in `apt` somehow.

**Reply by GarboMuffin** [reply]
I'm sorry but I don't think Scratch Addons should be worrying about your Linux distribution having a broken package manager. You can install node manually or pick your favorite from nvm, volta, fnm, n, etc. The npm version in apt is probably ancient so you'll want to do this anyways.

**Comment by GarboMuffin** [comment]
If `npm install` is finally going to become part of the development process here, it seems like it would be a great time to add type hints or type checking for large parts of Scratch internals with something like https://www.npmjs.com/package/@turbowarp/types. You don't need to migrate to TypeScript to use them.

I have an old proof of concept here: https://github.com/GarboMuffin/ScratchAddons/commits/types

We use something based on these in https://github.com/TurboWarp/extensions. It makes development significantly less painful.

Some examples of things that can be fully typed:

```js
  const vm = addon.tab.redux.state.scratchGui.vm;
  vm.runtime.on('targetWasCreated', target => {
    const listVar = target.lookupVariableByNameAndType('variable name', 'list');
    if (listVar) {
      listVar.value._monitorUpToDate = false;
    }
    vm.runtime.renderer.updateDrawableVisible(target.drawableID, false);
    const storage = vm.runtime.storage;
    storage.load(storage.AssetType.ImageBitmap, '...', storage.DataFormat.PNG)
      .then((asset) => asset.encodeDataURI());
    addon.tab.redux.dispatch({
      type: 'projectTitle/SET_PROJECT_TITLE',
      title: 'new title'
    });
  });
```

**Comment by Hans5958** [comment]
@TheColaber has something that has been written on Discord. I'm inviting him to join the discussion, either by adding to this post, or creating a new discussion post.

**Comment by TheColaber** [comment]
I guess I've been working a lot on creating a build step for sa, so I can tell you what I've learned in my exploration. I have developed a version of sa in rollup, implementing the following features: typescript, vue files, having a typed addon manifest using a typescript file, and a different method of loading addons without using comlink or something similar, and more features I don't remember. It went well but got messy. It was also a little slow, even though I hadn't implemented everything. Eventually I found that esbuild had a chance of working better, the only problem with esbuild is that not many people have given it as much love as they have for rollup. There are not many plugins created for it. I was using a chrome extension plugin for rollup that was working great, but that doesn't exist for esbuild. There was no chrome extension plugin created for it so I had to create one myself. Other than extra bundling code that we will need to write ourselves, esbuild works a lot better and faster. This is all testing and I'm sure if/when we do move to a build step, someone will find better ways to implement things. I don't have my esbuild version in a public repository yet, but when I do I will post it here.

**Reply by Ascor8522** [reply]
ESBuild is indeed a lot faster than Rollup and has more features and better out-of-the-box config.

About the plugin situation, I don't think such a plugin is necessary, and you should be able to bundle your code in the same way as any other project; independently if it's a browser extension or not. I'm always willing to help or give advice if needed.

Also, keep in mind that the benefits of using TypeScript and strongly typed language show when the complexity of the project increases. This requires a better code organization from the start, but imo the benefits largely outweight the small setup costs.
Anyway, I still  encourage you all to continue in this  direction, which I believe will help the project a lot, as much for the beginner (since you have a nicer API when developing) as for the more experienced members (a build step allows for finer control of the output).

**Reply by TheColaber** [reply]
> About the plugin situation, I don't think such a plugin is necessary, and you should be able to bundle your code in the same way as any other project; independently if it's a browser extension or not. I'm always willing to help or give advice if needed.

Yeah, you are right, It is not necessary, it doesn't need to be a plugin. I don't remember exactly why I made it a plugin in the first place... Maybe I did it so others could use it for their chrome extensions. The plugin is simple, it takes the manifest as the input and then gives rollup the js/ts files to bundle and copies the html files to the dist. I will make the code public and you can take a look and give me comments and tell me what I should change, because I'm not very experienced on this.

**Reply by TheColaber** [reply]
It's not complete, but here is my esbuild implementation: https://github.com/TheColaber/sa-esbuild

**Reply by TheColaber** [reply]
@Ascor8522, not sure if you saw

**Reply by Ascor8522** [reply]
@TheColaber had a quick look at it, will look more into it when I have more time

**Comment by DNin01** [comment]
Another option is [Deno](https://deno.com), which I've just started trying out.

**Reply by Hans5958** [reply]
Deno is *vastly* different than Node.js, and I don't think it is a good idea to have it here.
